app/src/main/kotlin/com/hippo/ehviewer/ui/main/RollingNumber.kt

```diff
@@ -0,0 +1,167 @@
+//    0 ~ 9 in Concept space is mapped to a circle rigidly
+//    null is mapped to a special point(0, - 1.528f * radius)
+//    keeping null - 0, null - 9, 0 - 9 have same distance
+// 3. Number Offset Space
+//    value is from -1 ~ 10, where -1 and 10 shows nothing
+//    radian of point in Intermediate space is mapped to here discontinuously
+
+// This controls animation speed
+private const val radius = 100f
+
+// Gaps between number in radian
+private const val gap = (2 * PI / 10).toFloat()
+private const val zeroDegree = -(PI / 2 - gap / 2).toFloat()
+
+// null is mapped to (0, - 1.528f * radius)
+private const val specialNodeDistanceFactor = 1.528f
+
+// Input: Concept space: 0 null 1 2 3 4 6 6 7 8 9
+// Output: Position in 2d euclidean space partial circle
+private fun conceptSpaceToIntermediate(value: Int?): Offset {
+    if (value != null) {
+        val theta = zeroDegree - value * gap
+        return Offset(cos(theta), sin(theta)) * radius
+    } else {
+        return Offset(0f, -specialNodeDistanceFactor) * radius
+    }
+}
+
+// Output: -1 ~ 10, where -1 and 10 shows nothing
+private fun intermediateToNumberOffset(circleOffset: Offset): Float {
+    val degree = atan2(circleOffset.y, circleOffset.x)
+    val normalized = normalize(zeroDegree - degree)
+    return if (normalized > 9 * gap) {
+        val reNormalized = normalized - 9 * gap - gap / 2
+        if (reNormalized > 0) {
+            // 0 side
+            reNormalized / gap * 2 - 1
+        } else {
+            // 9 side
+            10 + reNormalized / gap * 2
+        }
+    } else {
+        normalized / gap
+    }
+}
+
+// Normalize radian to [0, 2 * PI]
+private fun normalize(degree: Float) = (degree - 2 * PI * floor(degree / (2 * PI))).toFloat()
+
+// Convert 2d euclidean space partial circle to UI Offset
+private fun numberOffsetToUIOffset(size: IntSize, number: Float) = IntOffset(0, -(size.height * 2 * number).toInt())
+
+@Composable
+fun RollingNumber(number: Int, style: TextStyle = LocalTextStyle.current, width: Int? = null) {
+    val string = remember(number) { "$number" }
+    val transition = updateTransition(string)
+    val textMeasurer = rememberTextMeasurer()
+    val size by transition.animateIntSize { str ->
+        remember(str) { textMeasurer.measure(str, style).size }
+    }
+    val density = LocalDensity.current
+    val (styleNoSpacing, spacing) = remember(density, style) {
+        if (style.letterSpacing != TextUnit.Unspecified) {
+            val spacing = with(density) { style.letterSpacing.toDp() }
+            style.copy(letterSpacing = TextUnit.Unspecified) to spacing
+        } else {
+            style to 0.dp
+        }
+    }
+    Row(
+        modifier = Modifier.clipToBounds().layout { measurable, constraints ->
+            val placeable = measurable.measure(constraints)
+            layout(width = placeable.width, height = size.height) {
+                placeable.place(0, 0)
+            }
+        },
```

Comment:
`padding(horizontal = spacing)` as `spacedBy` does not include head and tail.
Rest looks good.

Reply:
Done